File: src/cardiocam/visao/pele.py
"""Segmentação de pele.

Dentro da caixa do rosto entra muita coisa que não é pele: cabelo, óculos,
sobrancelha, barba, fundo nos cantos. Esses pixels não pulsam, então só diluem o
sinal. Filtrar por cor de pele aumenta bastante a relação sinal-ruído sem custo
relevante.

Trabalhamos em YCrCb porque esse espaço separa luminância (Y) de crominância
(Cr, Cb). A cor da pele humana ocupa uma faixa estreita e surpreendentemente
estável em Cr e Cb, independentemente do tom: o que muda entre pessoas de pele
clara e escura é principalmente o Y. Aplicar o limiar só na crominância torna a
segmentação bem mais justa entre tons de pele do que limiarizar em RGB.
"""

from __future__ import annotations

import cv2
import numpy as np

# Faixa clássica de crominância da pele (Chai e Ngan, 1999).
CR_MINIMO, CR_MAXIMO = 133, 173
CB_MINIMO, CB_MAXIMO = 77, 127

# Descarta pixels queimados ou totalmente escuros, onde a informação de cor
# perde o sentido.
Y_MINIMO, Y_MAXIMO = 40, 250


def mascara_pele(
    imagem_bgr: np.ndarray,
    suavizar: bool = True,
    usar_luminancia: bool = True,
) -> np.ndarray:
    """Máscara booleana dos pixels classificados como pele.

    `suavizar` aplica abertura e fechamento morfológicos, que removem pixels
    isolados e fecham buracos pequenos, deixando regiões conexas.
    """
    if imagem_bgr is None or imagem_bgr.size == 0:
        return np.zeros((0, 0), dtype=bool)
    if imagem_bgr.ndim != 3 or imagem_bgr.shape[2] != 3:
        raise ValueError("A segmentação de pele espera uma imagem BGR de três canais.")

    ycrcb = cv2.cvtColor(imagem_bgr, cv2.COLOR_BGR2YCrCb)
    luminancia, cr, cb = cv2.split(ycrcb)

    dentro = (
        (cr >= CR_MINIMO) & (cr <= CR_MAXIMO) & (cb >= CB_MINIMO) & (cb <= CB_MAXIMO)
    )
    if usar_luminancia:
        dentro &= (luminancia >= Y_MINIMO) & (luminancia <= Y_MAXIMO)

    if not suavizar:
        return dentro

    binaria = dentro.astype(np.uint8)
    nucleo = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
    binaria = cv2.morphologyEx(binaria, cv2.MORPH_OPEN, nucleo)
    binaria = cv2.morphologyEx(binaria, cv2.MORPH_CLOSE, nucleo)
    return binaria.astype(bool)


def proporcao_de_pele(imagem_bgr: np.ndarray) -> float:
    """Fração de pixels de pele na imagem, de 0 a 1.

    Serve de indicador de qualidade: uma região de interesse com pouca pele
    provavelmente está desalinhada.
    """
    if imagem_bgr is None or imagem_bgr.size == 0:
        return 0.0
    mascara = mascara_pele(imagem_bgr)
    if mascara.size == 0:
        return 0.0
    return float(np.count_nonzero(mascara) / mascara.size)


def descartar_extremos(
    valores: np.ndarray, percentil: float = 5.0
) -> np.ndarray:
    """Remove as caudas da distribuição de intensidade.

    Reflexo especular (o brilho da tela na testa) e sombra dura são pixels que
    variam muito e não acompanham o pulso. Cortar os percentis extremos é uma
    forma barata de tirá-los da média.
    """
    if valores.size == 0:
        return valores
    if not 0.0 <= percentil < 50.0:
        raise ValueError("O percentil precisa estar entre 0 e 50.")
    if percentil == 0.0:
        return valores
    inferior = np.percentile(valores, percentil)
    superior = np.percentile(valores, 100.0 - percentil)
    dentro = (valores >= inferior) & (valores <= superior)
    return valores[dentro] if np.any(dentro) else valores

File: web/js/pele.js
/**
 * Segmentação de pele por crominância.
 *
 * Mesma ideia da versão em Python: converter para YCrCb e limiarizar apenas os
 * canais de cor, nunca o de brilho. A crominância da pele humana ocupa uma
 * faixa estreita e estável em qualquer tom, e o que varia entre pessoas de pele
 * clara e escura é sobretudo a luminância. Limiarizar só em Cr e Cb faz o
 * sistema medir todo mundo com a mesma competência, o que aqui é requisito de
 * correção e não detalhe de implementação.
 */

// Faixa clássica de crominância da pele (Chai e Ngan, 1999).
const CR_MINIMO = 133;
const CR_MAXIMO = 173;
const CB_MINIMO = 77;
const CB_MAXIMO = 127;

// Descarta pixels queimados ou totalmente escuros, onde a informação de cor
// perde o sentido.
const Y_MINIMO = 40;
const Y_MAXIMO = 250;

/** Decide se um pixel RGB é pele. */
export function classificarPele(r, g, b) {
  const y = 0.299 * r + 0.587 * g + 0.114 * b;
  if (y < Y_MINIMO || y > Y_MAXIMO) return false;
  const cr = (r - y) * 0.713 + 128;
  const cb = (b - y) * 0.564 + 128;
  return cr >= CR_MINIMO && cr <= CR_MAXIMO && cb >= CB_MINIMO && cb <= CB_MAXIMO;
}

/**
 * Média dos canais sobre os pixels de pele de uma região do canvas.
 *
 * A média espacial é o passo que viabiliza a medição inteira. A variação de
 * intensidade causada pelo pulso fica entre 0,1% e 1%, abaixo do ruído de
 * leitura de um pixel isolado; como esse ruído é aproximadamente independente
 * entre pixels, promediar N deles reduz o desvio por um fator de raiz de N.
 *
 * `passo` amostra um pixel a cada N para caber no orçamento de tempo de um
 * quadro em celular. Reduz o ganho de raiz de N, mas manter a taxa de quadros
 * estável importa mais.
 */
export function mediaDaPele(dados, largura, altura, passo = 2) {
  let somaR = 0;
  let somaG = 0;
  let somaB = 0;
  let pele = 0;
  let total = 0;

  for (let y = 0; y < altura; y += passo) {
    for (let x = 0; x < largura; x += passo) {
      const i = (y * largura + x) * 4;
      const r = dados[i];
      const g = dados[i + 1];
      const b = dados[i + 2];
      total++;
      if (classificarPele(r, g, b)) {
        somaR += r;
        somaG += g;
        somaB += b;
        pele++;
      }
    }
  }

  if (pele < 30) return null;
  return {
    vermelho: somaR / pele,
    verde: somaG / pele,
    azul: somaB / pele,
    pixels: pele,
    proporcao: total ? pele / total : 0,
  };
}
